windows/lcow_sandbox.go

```diff
@@ -0,0 +1,521 @@
+
+	// We don't do anything in safe mode unless the force flag has been passed, which
+	// is only the case for cleanup at driver termination.
+	if r.globalMode {
+		if !force {
+			log.G(ctx).Debugf("lcowdriver: terminateservicevm: %s (%s) - doing nothing as in global mode", id, context)
+			return nil
+		}
+		id = svmGlobalID
+	}
+
+	// Get the service VM and delete it from the map
+	svm, err := r.getServiceVM(ctx, id, true)
+	if err != nil {
+		return err
+	}
+
+	// We run the deletion of the scratch as a deferred function to at least attempt
+	// clean-up in case of errors.
+	defer func() {
+		if svm.scratchAttached {
+			scratchTargetFile := filepath.Join(r.root, scratchDirectory, fmt.Sprintf("%s.vhdx", id))
+			log.G(ctx).Debugf("lcowdriver: terminateservicevm: %s (%s) - deleting scratch %s", id, context, scratchTargetFile)
+			if err := os.Remove(scratchTargetFile); err != nil {
+				log.G(ctx).Warnf("failed to remove scratch file %s (%s): %s", scratchTargetFile, context, err)
+			}
+		}
+	}()
+
+	// Nothing to do if it's not running
+	if svm.config.Uvm != nil {
+		log.G(ctx).Debugf("lcowdriver: terminateservicevm: %s (%s) - calling terminate", id, context)
+		if err := svm.config.Uvm.Terminate(); err != nil {
+			return fmt.Errorf("failed to terminate utility VM (%s): %s", context, err)
+		}
+
+		log.G(ctx).Debugf("lcowdriver: terminateservicevm: %s (%s) - waiting for utility VM to terminate", id, context)
+		if err := svm.config.Uvm.WaitTimeout(time.Duration(svm.config.UvmTimeoutSeconds) * time.Second); err != nil {
+			return fmt.Errorf("failed waiting for utility VM to terminate (%s): %s", context, err)
+		}
+	}
+
+	log.G(ctx).Debugf("lcowdriver: terminateservicevm: %s (%s) - success", id, context)
+	return nil
+}
+
+// CreateReadWrite creates a layer that is writable for use as a container
+// file system. That equates to creating a sandbox.
+func (r *LCOWRuntime) CreateReadWrite(ctx context.Context, id, parent string) error {
+	title := fmt.Sprintf("lcowdriver: createreadwrite: id %s", id)
+	log.G(ctx).Debugf(title)
+
+	// First we need to create the folder
+	if err := r.CreateLayerFolder(id, parent); err != nil {
+		return err
+	}
+
+	// Look for an explicit sandbox size option.
+	sandboxSize := uint64(client.DefaultVhdxSizeGB)
+
+	// Massive perf optimisation here. If we know that the RW layer is the default size,
+	// and that the cached sandbox already exists, and we are running in safe mode, we
+	// can just do a simple copy into the layers sandbox file without needing to start a
+	// unique service VM. For a global service VM, it doesn't really matter. Of course,
+	// this is only the case where the sandbox is the default size.
+	//
+	// Make sure we have the sandbox mutex taken while we are examining it.
+	if sandboxSize == client.DefaultVhdxSizeGB {
+		log.G(ctx).Debugf("%s: locking cachedSandboxMutex", title)
+		r.cachedSandboxMutex.Lock()
+		_, err := os.Stat(r.cachedSandboxFile)
```

Comment:
same as for the scratch vhd, just create it when the plugin is initialized. That way the mutex can go away.